index.go

```diff
@@ -1,47 +1,2085 @@
+					img = mutate.Subject(img, *upsertSubject).(v1.Image)
+				}
+
+				iMap.Store(img, ops)
+				return nil
+			})
+
+			if err = errGroup.Wait(); err != nil {
+				return err
+			}
+		default:
+			return ErrUnknownMediaType
+		}
+	}
+
+	wg.Wait()
+	i.Annotate = Annotate{}
+	iMap.Range(func(key, value any) bool {
+		switch v := key.(type) {
+		case v1.Image:
+			ops, ok := value.([]layout.Option)
+			if !ok {
+				return false
+			}
+
+			err = path.AppendImage(v, ops...)
+			if err != nil {
+				errs.Errors = append(errs.Errors, SaveDiagnostic{
+					Cause: err,
+				})
+			}
+			return true
+		case v1.ImageIndex:
+			ops, ok := value.([]layout.Option)
+			if !ok {
+				return false
+			}
+
+			err = path.AppendIndex(v, ops...)
+			if err != nil {
+				errs.Errors = append(errs.Errors, SaveDiagnostic{
+					Cause: err,
+				})
+			}
+			return true
+		default:
+			return false
+		}
+	})
+
+	if len(errs.Errors) != 0 {
+		return errs
+	}
+
+	var removeHashes = make([]v1.Hash, 0)
+	for _, h := range i.RemovedManifests {
+		if _, ok := i.Images[h]; !ok {
+			removeHashes = append(removeHashes, h)
+			delete(i.Images, h)
+		}
+	}
+
+	err = path.RemoveDescriptors(match.Digests(removeHashes...))
+	if err != nil {
+		return err
+	}
+
+	i.RemovedManifests = make([]v1.Hash, 0)
+	return nil
+}
+
+func (i *Index) Push(ops ...IndexPushOption) error {
+	var pushOps = &PushOptions{}
+
+	if len(i.RemovedManifests) != 0 || len(i.Annotate.Instance) != 0 {
+		if err := i.Save(); err != nil {
+			return err
+		}
+	}
+
+	for _, op := range ops {
+		err := op(pushOps)
+		if err != nil {
+			return err
+		}
+	}
+
+	ref, err := name.ParseReference(
+		i.Options.Reponame,
+		name.WeakValidation,
+		name.Insecure,
+	)
+	if err != nil {
+		return err
+	}
+
+	layoutPath := filepath.Join(i.Options.XdgPath, i.Options.Reponame)
+	path, err := layout.FromPath(layoutPath)
+	if err != nil {
+		return err
+	}
+
+	imageIndex, err := path.ImageIndex()
+	if err != nil {
+		return err
+	}
+
+	if pushOps.Format != "" {
+		mfest, err := i.IndexManifest()
+		if err != nil {
+			return err
+		}
+
+		if mfest == nil {
+			return ErrManifestUndefined
+		}
+
+		if pushOps.Format != types.MediaType("") && pushOps.Format != mfest.MediaType {
+			imageIndex = mutate.IndexMediaType(imageIndex, pushOps.Format)
+		}
 	}
+
+	err = remote.WriteIndex(
+		ref,
+		imageIndex,
+		remote.WithAuthFromKeychain(i.Options.KeyChain),
+		remote.WithTransport(getTransport(pushOps.Insecure)),
+	)
+	if err != nil {
+		return err
+	}
+
+	if pushOps.Purge {
+		return i.Delete()
+	}
+
+	return nil
+}
+
+func (i *Index) Inspect() error {
+	mfest, err := i.IndexManifest()
+	if err != nil {
+		return err
+	}
+
+	if mfest == nil {
+		return ErrManifestUndefined
+	}
+
+	if len(i.RemovedManifests) != 0 || len(i.Annotate.Instance) != 0 {
+		return ErrIndexNeedToBeSaved
+	}
+
+	mfestBytes, err := json.MarshalIndent(mfest, "", "		")
+	if err != nil {
+		return err
+	}
+
+	return errors.New(string(mfestBytes))
```

Comment:
I think the signature for this method is wrong, if we want to return the bytes. probably should return (string, error). 

the output in pack is really weird.

<img width="1482" alt="Screenshot 2024-02-09 at 9 36 03 AM" src="https://github.com/husni-faiz/imgutil/assets/1181799/d81ec65d-aa1e-449b-9863-f7c552517c82">

look it is being printed as **ERROR**, because of that it is not easy to pipe to other tooling like `jq`